services/tracing.py

```diff
@@ -120,3 +123,30 @@ def wrap_with_span(*args, **kwargs):
         return span_decorator
     else:
         return span_decorator(_func)
+
+
+def _get_argument_from_function_signature(
+    func_signature, func_bound_args, func_kwargs, parameter_name
+):
+    """
+    Find the value of a named parameter passed to a function from (in order):
+    1) the function kwargs directly; if a function signature takes a parameter
+       like `**kwargs`, we can retrieve a named parameter from the keyword arguments
+       there
+    2) the bound args retrieved from the function signature; this will find any
+       explicity passed values when the function was called.
+    3) the parameter default value, if there is one
+    4) Finally, raises an exception if we can't find a value for the expected parameter
+    """
+    if parameter_name in func_kwargs:
+        return func_kwargs[parameter_name]
+    if parameter_name in func_bound_args:
+        return func_bound_args[parameter_name]
+    # an argument for this parameter was not passed explicitly in the function call;
+    # look for a default value
+    param = func_signature.parameters.get(parameter_name)
```

Comment:
Can we extract default parameters in the decorator body at import time into a dict, and then just look them up at call time?

Combined with pulling out the signature in the decorator at import time, we can build the default param dict then too.

If so, this function might be small enough to not bother with, we can just inline it in the loop, as we can just use ChainMap to look up the parameter in a series of dicts:

(Note, this also include removing the need for _set_attributes, as per my other comment)

```
...
from collections import ChainMap
...


def instrument(...):
    ...

    def span_decorator(func):
        ...
        default_args = {
             k: v.default for k, v in signature(func).parameters.items()
             if v.default != inspect.Parameter.empty,
        }
        bound_args = {}
     
        @wraps(func)
        def wrap_with_span(*args, **kwargs):
            ...
            if func_attributes is not None:
                bound_args = func_signature.bind(*args, **kwargs).arguments
  
            param_lookup = ChainMap(kwargs, bound_args, default_args)
            for k, v in func_attributes.items():
                try:
                    attributes_dict[k] = str(param_lookup[v])
                except ...:

            with tracer.start_as_current_span(
                name, record_exception=record_exception, attributes=attribute_dict,
            ):
                return func(*args, **kwargs)

        return wrap_with_span
 
```